File: 01-基础篇/code/day4/api_demo.py
"""
第4天：API 基础调用演示
"""
import os
import json
import time

API_KEY = os.environ.get("DEEPSEEK_API_KEY")
if not API_KEY:
    env_path = os.path.expanduser("~/.hermes/.env")
    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                if "DEEPSEEK_API_KEY" in line:
                    API_KEY = line.split("=", 1)[1].strip().strip('"').strip("'")
                    break

import requests


def call_api(messages, temperature=0.7, max_tokens=1024, system=None):
    """调用 DeepSeek API 并返回完整响应"""
    if system:
        messages = [{"role": "system", "content": system}] + messages

    resp = requests.post(
        "https://api.deepseek.com/chat/completions",
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json"
        },
        json={
            "model": "deepseek-chat",
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens
        },
        timeout=30
    )
    return resp.json()


def call_with_retry(messages, temperature=0, max_retries=3):
    """带重试的 API 调用"""
    for attempt in range(max_retries):
        try:
            resp = call_api(messages, temperature=temperature)
            if "error" not in resp:
                return resp
            if attempt < max_retries - 1:
                wait = 2 ** attempt
                print(f"  [限流] {wait}s 后重试...")
                time.sleep(wait)
        except Exception as e:
            if attempt < max_retries - 1:
                wait = 2 ** attempt
                print(f"  [错误] {e}，{wait}s 后重试...")
                time.sleep(wait)
            else:
                raise e
    return resp


# ========== 实验1：查看完整响应结构 ==========
print("=" * 60)
print("实验1：完整响应结构")
print("=" * 60)

response = call_api(
    [{"role": "user", "content": "2 + 3 × 4 等于多少？"}],
    temperature=0
)

print(f"\n响应 ID: {response.get('id')}")
print(f"模型: {response.get('model')}")
print(f"回答: {response['choices'][0]['message']['content']}")
print(f"停止原因: {response['choices'][0]['finish_reason']}")
usage = response.get('usage', {})
print(f"\nToken 用量:")
print(f"  输入: {usage.get('prompt_tokens')} tokens")
print(f"  输出: {usage.get('completion_tokens')} tokens")
print(f"  总计: {usage.get('total_tokens')} tokens")

time.sleep(2)

# ========== 实验2：Temperature 对比 ==========
print("\n" + "=" * 60)
print("实验2：Temperature 对比 (0 vs 1.5)")
print("=" * 60)

prompt = "给我一个 Python 函数名建议（只要名字）："

r_cold = call_api([{"role": "user", "content": prompt}], temperature=0)
r_hot = call_api([{"role": "user", "content": prompt}], temperature=1.5)

print(f"\nTemperature=0:   {r_cold['choices'][0]['message']['content'].strip()}")
print(f"Temperature=1.5: {r_hot['choices'][0]['message']['content'].strip()}")

time.sleep(2)

# ========== 实验3：多轮对话 ==========
print("\n" + "=" * 60)
print("实验3：多轮对话")
print("=" * 60)

messages = [
    {"role": "user", "content": "我的名字是小明。"},
]
r1 = call_api(messages, temperature=0)
print(f"用户: 我的名字是小明。")
print(f"助手: {r1['choices'][0]['message']['content']}\n")

messages.append(r1['choices'][0]['message'])
messages.append({"role": "user", "content": "我刚才说我叫什么名字？"})

r2 = call_api(messages, temperature=0)
print(f"用户: 我刚才说我叫什么名字？")
print(f"助手: {r2['choices'][0]['message']['content']}")

File: 01-基础篇/code/day4/ApiDemo.java
package ai.learning.day4;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpRequest.BodyPublishers;

/**
 * 第4天：API 基础调用演示
 *
 * 运行前设置环境变量：
 * export DEEPSEEK_API_KEY=$(grep DEEPSEEK_API_KEY ~/.hermes/.env | cut -d= -f2 | tr -d '"')
 *
 * 编译：javac ApiDemo.java
 * 运行：java ai.learning.day4.ApiDemo
 */
public class ApiDemo {

    static final HttpClient client = HttpClient.newHttpClient();

    static String callApi(String jsonBody) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create("https://api.deepseek.com/chat/completions"))
            .header("Authorization", "Bearer " + System.getenv("DEEPSEEK_API_KEY"))
            .header("Content-Type", "application/json")
            .POST(BodyPublishers.ofString(jsonBody))
            .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return response.body();
    }

    static String extractContent(String json) {
        int idx = json.indexOf("\"content\":\"");
        if (idx < 0) return "解析失败";
        int start = idx + 11;
        StringBuilder sb = new StringBuilder();
        for (int i = start; i < json.length(); i++) {
            char c = json.charAt(i);
            if (c == '\\') {
                if (i + 1 < json.length()) {
                    char n = json.charAt(i + 1);
                    if (n == 'n') { sb.append('\n'); i++; }
                    else if (n == '"') { sb.append('"'); i++; }
                    else if (n == '\\') { sb.append('\\'); i++; }
                    else { sb.append(c); }
                }
            } else if (c == '"') {
                break;
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    static void printUsage(String json) {
        int pt = json.indexOf("\"prompt_tokens\":");
        int ct = json.indexOf("\"completion_tokens\":");
        int tt = json.indexOf("\"total_tokens\":");
        if (pt > 0) System.out.println("  输入: " + extractNum(json, pt));
        if (ct > 0) System.out.println("  输出: " + extractNum(json, ct));
        if (tt > 0) System.out.println("  总计: " + extractNum(json, tt));
    }

    static String extractNum(String json, int start) {
        int i = json.indexOf(':', start) + 1;
        StringBuilder sb = new StringBuilder();
        while (i < json.length() && Character.isDigit(json.charAt(i))) {
            sb.append(json.charAt(i++));
        }
        return sb.toString();
    }

    static String makeBody(String userMsg, double temp, int maxTokens) {
        return String.format("""
            {
                "model": "deepseek-chat",
                "messages": [{"role": "user", "content": "%s"}],
                "temperature": %.1f,
                "max_tokens": %d
            }
            """, escapeJson(userMsg), temp, maxTokens);
    }

    static String escapeJson(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n");
    }

    public static void main(String[] args) throws Exception {
        // 实验1
        System.out.println("=".repeat(60));
        System.out.println("实验1：Token 用量");
        System.out.println("=".repeat(60));

        String body1 = makeBody("请用20字以内介绍 Python。", 0.7, 200);
        String resp1 = callApi(body1);
        System.out.println("回答: " + extractContent(resp1));
        printUsage(resp1);

        // 实验2
        System.out.println("\n" + "=".repeat(60));
        System.out.println("实验2：Temperature 对比");
        System.out.println("=".repeat(60));

        String cold = callApi(makeBody("说一个数字（只输出数字）：", 0, 50));
        String hot = callApi(makeBody("说一个数字（只输出数字）：", 1.5, 50));
        System.out.println("Temperature=0:   " + extractContent(cold));
        System.out.println("Temperature=1.5: " + extractContent(hot));
    }
}
